File: ios/JustSpentTests/Utilities/SharedTestDataLoader.swift
//
//  SharedTestDataLoader.swift
//  JustSpentTests
//
//  Loader for shared test data from JSON files
//  Loads test data from shared/test-data/ directory
//

import Foundation

/// Loader for shared test data
class SharedTestDataLoader {

    // MARK: - Voice Test Data Structures

    struct VoiceTestData: Codable {
        let version: String
        let description: String
        let test_suites: TestSuites
    }

    struct TestSuites: Codable {
        let currency_detection: TestSuite
        let amount_extraction: TestSuite
        let written_numbers: TestSuite
        let real_world_scenarios: TestSuite
        let edge_cases: TestSuite
        let symbol_normalization: TestSuite
        let disambiguation: TestSuite
    }

    struct TestSuite: Codable {
        let description: String
        let tests: [VoiceTestCase]
    }

    struct VoiceTestCase: Codable {
        let id: String
        let input: String
        let expected_currency: String?
        let expected_amount: Double?
        let expected_category: String?
        let expected_merchant: String?
        let description: String
        let default_currency: String?
        let normalized_output: String?
        let expected_behavior: String?
        let note: String?
    }

    // MARK: - Loading Methods

    /// Load voice test data from JSON
    static func loadVoiceTestData() throws -> VoiceTestData {
        // Try multiple possible paths for the JSON file
        let possiblePaths = [
            // Path 1: Relative from current directory
            "../../shared/test-data/voice-test-data.json",
            // Path 2: Relative from project root
            "../../../shared/test-data/voice-test-data.json",
            // Path 3: Absolute from FileManager current directory
            "\(FileManager.default.currentDirectoryPath)/../../shared/test-data/voice-test-data.json",
            // Path 4: From source root (for CI environments)
            "/home/user/just-spent/shared/test-data/voice-test-data.json",
            // Path 5: From runner work directory (GitHub Actions)
            "/home/runner/work/just-spent/just-spent/shared/test-data/voice-test-data.json"
        ]

        var lastError: Error?
        for path in possiblePaths {
            if let data = try? Data(contentsOf: URL(fileURLWithPath: path)) {
                let decoder = JSONDecoder()
                return try decoder.decode(VoiceTestData.self, from: data)
            }
        }

        // If none of the paths worked, throw error with all attempted paths
        let attemptedPaths = possiblePaths.joined(separator: "\n  - ")
        throw TestDataError.fileNotFound("Attempted paths:\n  - \(attemptedPaths)")
    }

    /// Get currency detection tests
    static func getCurrencyDetectionTests() throws -> [VoiceTestCase] {
        let data = try loadVoiceTestData()
        return data.test_suites.currency_detection.tests
    }

    /// Get amount extraction tests
    static func getAmountExtractionTests() throws -> [VoiceTestCase] {
        let data = try loadVoiceTestData()
        return data.test_suites.amount_extraction.tests
    }

    /// Get written number tests
    static func getWrittenNumberTests() throws -> [VoiceTestCase] {
        let data = try loadVoiceTestData()
        return data.test_suites.written_numbers.tests
    }

    /// Get edge case tests
    static func getEdgeCaseTests() throws -> [VoiceTestCase] {
        let data = try loadVoiceTestData()
        return data.test_suites.edge_cases.tests
    }

    /// Get real world scenario tests
    static func getRealWorldTests() throws -> [VoiceTestCase] {
        let data = try loadVoiceTestData()
        return data.test_suites.real_world_scenarios.tests
    }

    /// Get disambiguation tests
    static func getDisambiguationTests() throws -> [VoiceTestCase] {
        let data = try loadVoiceTestData()
        return data.test_suites.disambiguation.tests
    }
}

// MARK: - Errors

enum TestDataError: Error {
    case fileNotFound(String)
    case decodingFailed(Error)

    var localizedDescription: String {
        switch self {
        case .fileNotFound(let path):
            return "Test data file not found at: \(path)"
        case .decodingFailed(let error):
            return "Failed to decode test data: \(error.localizedDescription)"
        }
    }
}

File: android/app/src/test/java/com/justspent/expense/utils/SharedTestDataLoader.kt
package com.justspent.expense.utils

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.decodeFromString
import java.io.File

/**
 * Loader for shared test data from JSON files
 * Loads test data from shared/test-data/ directory
 */
object SharedTestDataLoader {

    private val json = Json { ignoreUnknownKeys = true }

    /**
     * Voice test data structure
     */
    @Serializable
    data class VoiceTestData(
        val version: String,
        val description: String,
        val test_suites: TestSuites
    )

    @Serializable
    data class TestSuites(
        val currency_detection: TestSuite,
        val amount_extraction: TestSuite,
        val written_numbers: TestSuite,
        val real_world_scenarios: TestSuite,
        val edge_cases: TestSuite,
        val symbol_normalization: TestSuite,
        val disambiguation: TestSuite
    )

    @Serializable
    data class TestSuite(
        val description: String,
        val tests: List<VoiceTestCase>
    )

    @Serializable
    data class VoiceTestCase(
        val id: String,
        val input: String,
        val expected_currency: String? = null,
        val expected_amount: Double? = null,
        val expected_category: String? = null,
        val expected_merchant: String? = null,
        val description: String,
        val default_currency: String? = null,
        val normalized_output: String? = null,
        val expected_behavior: String? = null,
        val note: String? = null
    )

    /**
     * Load voice test data
     */
    fun loadVoiceTestData(): VoiceTestData {
        val jsonFile = File("../../shared/test-data/voice-test-data.json")
        require(jsonFile.exists()) { "Voice test data file not found: ${jsonFile.absolutePath}" }

        val jsonContent = jsonFile.readText()
        return json.decodeFromString<VoiceTestData>(jsonContent)
    }

    /**
     * Get specific test suite
     */
    fun getCurrencyDetectionTests(): List<VoiceTestCase> {
        return loadVoiceTestData().test_suites.currency_detection.tests
    }

    fun getAmountExtractionTests(): List<VoiceTestCase> {
        return loadVoiceTestData().test_suites.amount_extraction.tests
    }

    fun getWrittenNumberTests(): List<VoiceTestCase> {
        return loadVoiceTestData().test_suites.written_numbers.tests
    }

    fun getEdgeCaseTests(): List<VoiceTestCase> {
        return loadVoiceTestData().test_suites.edge_cases.tests
    }

    fun getRealWorldTests(): List<VoiceTestCase> {
        return loadVoiceTestData().test_suites.real_world_scenarios.tests
    }

    fun getDisambiguationTests(): List<VoiceTestCase> {
        return loadVoiceTestData().test_suites.disambiguation.tests
    }
}
